pkg/reconciler/tenancy/defaultapibindinglifecycle/default_apibinding_lifecycle_reconcile.go

```diff
@@ -0,0 +1,214 @@
+
+			logger := logger.WithValues("apiExport.path", exportRef.Path, "apiExport.name", exportRef.Export)
+			ctx := klog.NewContext(ctx, logger)
+
+			apiBindingName := initialization.GenerateAPIBindingName(clusterName, exportRef.Path, exportRef.Export)
+			logger = logger.WithValues("apiBindingName", apiBindingName)
+
+			// if _, err = c.getAPIBinding(clusterName, apiBindingName); err == nil {
+			// 	logger.V(4).Info("APIBinding already exists - skipping creation")
+			// 	continue
+			// }
+
+			apiBinding := &apisv1alpha1.APIBinding{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: apiBindingName,
+				},
+				Spec: apisv1alpha1.APIBindingSpec{
+					Reference: apisv1alpha1.BindingReference{
+						Export: &apisv1alpha1.ExportBindingReference{
+							Path: exportRef.Path,
+							Name: apiExport.Name,
+						},
+					},
+				},
+			}
+
+			for _, exportClaim := range apiExport.Spec.PermissionClaims {
+				// For now we automatically accept DefaultAPIBindings
+				acceptedClaim := apisv1alpha1.AcceptablePermissionClaim{
+					PermissionClaim: exportClaim,
+					State:           apisv1alpha1.ClaimAccepted,
+				}
+
+				apiBinding.Spec.PermissionClaims = append(apiBinding.Spec.PermissionClaims, acceptedClaim)
+			}
+
+			logger = logging.WithObject(logger, apiBinding)
+
+			existingBinding, err := c.getAPIBinding(clusterName, apiBinding.Name)
+			if err != nil {
+				if !apierrors.IsNotFound(err) {
+					errors = append(errors, err)
+					continue
+				}
+				if _, err := c.createAPIBinding(ctx, clusterName.Path(), apiBinding); err != nil {
+					errors = append(errors, err)
+					continue
+				}
+				logger.V(2).Info("created APIBinding")
+			} else {
+				// TODO: Overwrite everything?
+				apiBinding.SetResourceVersion(existingBinding.ResourceVersion)
+				if _, err := c.updateAPIBinding(ctx, clusterName.Path(), apiBinding); err != nil {
+					errors = append(errors, err)
+					continue
+				}
+				logger.V(2).Info("updated APIBinding")
+			}
+		}
+	}
+
+	if len(errors) > 0 {
+		logger.Error(utilerrors.NewAggregate(errors), "error reconciling APIBindings")
+
+		if someExportsMissing {
+			// Retry if any APIExports are missing, as it's possible they'll show up (cache server slow to catch up,
+			// arrive via replication)
+			return utilerrors.NewAggregate(errors)
+		}
+
+		return nil
+	}
+
+	// Make sure all the expected bindings are there & ready to use
+	var incomplete []string
+
+	for exportRef := range requiredExportRefs {
+		binding, exists := exportToBinding[apisv1alpha1.ExportBindingReference{
+			Path: exportRef.Path,
+			Name: exportRef.Export,
+		}]
+		if !exists {
+			incomplete = append(incomplete, fmt.Sprintf("for APIExport %s|%s", exportRef.Path, exportRef.Export))
+			continue
+		}
+
+		if !conditions.IsTrue(binding, apisv1alpha1.InitialBindingCompleted) {
+			incomplete = append(incomplete, binding.Name)
+		}
+	}
+
+	if len(incomplete) > 0 {
+		sort.Strings(incomplete)
+		return nil
+	}
```

Comment:
I think there is something missing here, what are we doing with `incomplete`? Should we log/surface this in any way?

Reply:
Yes. I added the setting of the conditions.